sdk/dotnet/Extensions.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright 2016-2019, Pulumi Corporation.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace Pulumi.Docker
+{
+    internal static class UnwrapExtensions
+    {
+        public static Input<ImmutableArray<T>?> Unwrap<T>(this InputList<T>? items)
+        {
+            return items == null ? Output.Create((ImmutableArray<T>?)null) : items.Apply(v => (ImmutableArray<T>?)v);
+        }
+
+        public static Input<T?> Unwrap<T>(this Input<T>? input) where T : class
+        {
+            return input == null ? Output.Create((T?)null) : input.Apply(v => (T?)v);
+        }
+
+        public static Input<CacheFromUnwrap?> Unwrap(this InputUnion<bool, CacheFrom>? build)
+        {
+            if (build == null)
+                return Output.Create((CacheFromUnwrap?)null);
+
+            return build
+                .Apply(v => v.IsT0 ? new ImmutableArray<string>() : v.AsT1.Stages.Unwrap())
+                .Apply(v => v != null ? new CacheFromUnwrap { Stages = v } : null);
+        }
+
+        public static Input<Union<string, DockerBuildUnwrap>> Unwrap(this InputUnion<string, DockerBuild> build)
```

Comment:
these could be represented with a lower layer extension that operates on an `InputUnion<X, Y>` and which takes some helper functions to map from an X to an UnwrapX.